### ISSUE TITLE
Dangerous behaviour of `R.add`

### ISSUE THREAD

**Issue by dmitriz** [ramda/ramda#2037]
Right now `add` silently ignores all of its arguments starting from the 3rd:

```js
add(2,3,10) // => 6
```

I find this behaviour somewhat dangerous that may lead to hard to detect bugs. 

I have heard the reasoning that `add` is curried but would still see a logical way to solve this problem. Note that even currently `add(3)` is a function `Number -> Number` and is never confused with the binary `add(x,y)` with optional 2nd argument, which would make `add(3)` a `Number`. So, despite a potential confusion, there actually isn't one.

On the other hand, `add(x,y,z)` does actually behave as having `z` optional, which does not seem consistent with `add(x,y)`.  Is there any chance/interest to improve it?

Fundamentally, `add` is overloaded. 

Question: What is the signature of `add`?
Answer: It depends on the usage context:

```js
// (Number, Number) -> Number
add(x,y)

// Number -> Number -> Number
add(x)(y)
```

But nothing prevents us from extending this overloading to more variables exactly the same way. (I am still using `add` but it can also be named something else like `addV` meaning variadic, whereas `addN` can mean fixed arity like `curryN`):

```js
// (Number, Number, Number) -> Number
add(x,y,z)

// (Number, Number) -> Number -> Number
add(x,y)(z)

// Number -> Number -> Number -> Number
add(x)(y)(z)
```

Question. So what is `add(3)`? 
Answer. Again and exactly the same way, it depends on the context:

```js
// Number
add(3)

// Number -> Number
add(3)(x)

// (Number, Number) -> Number
add(3)(x,y)

// Number -> Number -> Number
add(3)(x)(y)
```

Q. Ok, so how do I know when `add(3)` is a `Number` and when it is a function?
A. Let us look at real examples:

```js
// One-argument function, would you ever confuse this for Number?
[1,2,3].map(add(3)) / => [4,5,6]

// Any room for confusion?
[1,2,3].map(add(3, 1)) / => [5,6,7]
// currently error: '4 is not a function'

// so we have to resort to the more verbose
[1,2,3].map(add(3), add(1)) / => [5,6,7]
// which actually works even if not respecting its signature
// Functor f => (a → b) → f a → f b
```

But what if `add` acts as curried 4-adic, then

```js
// add(3,1) :: Number -> Number -> Number
[1,2,3].map(add(3, 1)) // => [x=>x+5, x=>x+6, x=>x+7] ???
```

Ok, so we need to agree that `add` is "lazy" and uses the minimum arity.
That would remove any confusion.

And if we want to fix the arity, there is always `addN` ;)

**Comment by CrossEye** [comment]
This is a long post, and I can't tell if it there is any actual problem being reported or any change being suggested.

> Right now `add` silently ignores all of its arguments starting from the 3rd:
> 
> ```js
> add(2,3,10) // => 6
> ```

**5**, I would hope!

The behavior of `add` is very much consistent with almost all functions in Ramda.  We have (almost) no variadic functions.  All functions have a fixed arity.  `add` has arity 2.

> [...] Fundamentally, add is overloaded.

Yes, all Ramda functions of more than a single argument are like this.

```js
// add :: (Number, Number) -> Number
// add :: Number -> (Number -> Number)
```

Neither formulation has any particular precedence, although the former is the underlying implementation and the latter is created by our currying.   


> Q. Ok, so how do I know when `add(3)` is a `Number` and when it is a function?
> A. Let us look at real examples:
> 
> ```js
> // One-argument function, would you ever confuse this for Number?
> [1,2,3].map(add(3)) / => [4,5,6]


> ```

So far,so good.

> ```js
> // Any room for confusion?
> [1,2,3].map(add(3, 1)) / => [5,6,7]
> // currently error: '4 is not a function'
> ```

Well no, no room for confusion, but by the rules of Javascript it cannot work like you seem to want.  Javascript is eagerly evaluated. So when you pass an expression to `map`, it must be evaluated before `map` is called.  So the first step is to evaluated `add(3, 1)`, which will yield 4, and then call `[1, 2, 3].map(4)`, and hence the error you see.  That expression cannot know its own context.  You cannot say to `add(3, 1)`, "Hey, you're about to be called in a `map`, so please return a function rather than a scalar value."

> ```js
> // so we have to resort to the more verbose
> [1,2,3].map(add(3), add(1)) / => [5,6,7]
> // which actually works even if not respecting its signature
> // Functor f => (a → b) → f a → f b
> ```

Two things.  First, that signature is for Ramda's `map` function and not the `map` method on `Array.prototype`.  It's equivalent, for arrays, though, to 

```js
// map :: (a-> b) -> [a] -> [b]
```

Second, [it doesn't do what you want][r1]:

```js
[1,2,3].map(add(3), add(1)); //=> [4, 5, 6]
```

`Array.prototype.map` simply ignores the second parameter.

There is an odd way in which this is related to something you might want:

```js
R.map(add(3), add(1))
```

But this has only to do with the fact that `map` for functions turns out to be functional composition, and hence the above is very much equivalent to `add(4)`.

And you conclude with this:

> Ok, so we need to agree that add is "lazy" and uses the minimum arity.
> That would remove any confusion.

If this is you in the end reaching the conclusion that Ramda has done the right thing, that's fine.  This might then make an interesting Gist or blog post, but perhaps not an issue for the Ramda list.  If that's the case, please close this issue.

If you do think there is something for Ramda to change, could you please state it succinctly?  

Feel free to respond in either case; I just can't quite tell what you're looking for.

  [r1]: https://goo.gl/CU7cH3

**Comment by dmitriz** [comment]
> This is a long post, and I can't tell if it there is any actual problem being reported or any change being suggested.

Sorry :). The problem is actually in the title and is stated in the first paragraph.
Using generic name like `add`, which means variadic for most people,
for something different, i.e. binary addition, is confusing and potentially dangerous.
Especially for less experienced newcomers.

**Scroll down to the list of concrete suggestions! **

>>    add(2,3,10) // => 6

> 5, I would hope!

Yes, sorry ;)


> The behavior of add is very much consistent with almost all functions in Ramda. We have (almost) no variadic functions. All functions have a fixed arity. add has arity 2.

Variadic functions are commonly used in JavaScript.
And even in Ramda I can see  `compose` and `pipe` :)
Specifically `add` is easy to confuse for variadic
for anyone less familiar with Ramda's intricacies.

**Suggestion**

Use `add2` and `add3` instead of `add`
for corresponding binary and ternary versions.
It is already done in Sanctuary, for instance:
https://github.com/sanctuary-js/sanctuary#lift2--applyf--a-b-c---fa---fb---fc

 ```js
// Possible changed behaviour - Any room for confusion?
[1,2,3].map(add(3, 1)) / => [5,6,7]
// currently error: '4 is not a function'
```

> Well no, no room for confusion, but by the rules of Javascript it cannot work like you seem to want. Javascript is eagerly evaluated. So when you pass an expression to map, it must be evaluated before map is called. So the first step is to evaluated add(3, 1), which will yield 4, and then call [1, 2, 3].map(4), and hence the error you see. That expression cannot know its own context. You cannot say to add(3, 1), "Hey, you're about to be called in a map, so please return a function rather than a scalar value."

I see, but can I not evaluate `add(3,1)` to both number and function (stored in an object, and lazily for performance reasons)? Now `map(...)` comes along looking at this object and says "Hey, I can pick the function variant" and off you go ;) Intuitive for the user and no errors. Does it make sense?

```js
// so we have to resort to the more verbose
[1,2,3].map(add(3), add(1)) / => [5,6,7]
// which actually works even if not respecting its signature
// Functor f => (a → b) → f a → f b
```
> Two things. First, that signature is for Ramda's map function and not the map method on Array.prototype. 

Good point, shows the danger of identical names 😈  
So with Ramda we have:

```js
R.map(add(3), add(3))([1]) 
// => ... "331" ... horror? ;)
```

>Second, it doesn't do what you want:

    [1,2,3].map(add(3), add(1)); //=> [4, 5, 6]

> Array.prototype.map simply ignores the second parameter.

Ah yes, you are right! Easy to confuse but as `map` is unary, actually makes sense.


>There is an odd way in which this is related to something you might want:

    R.map(add(3), add(1))

>But this has only to do with the fact that map for functions turns out to be functional composition, and hence the above is very much equivalent to add(4).

Hm... I actually get this:

```js
R.map(add(3), add(1))([1,2,3]) 
// => "311,2,3"
```
http://ramdajs.com/repl/?v=0.19.0#?R.map%28add%283%29%2C%20add%283%29%29%28%5B1%5D%29%0AR.map%28add%283%29%2C%20add%281%29%29%28%5B1%2C2%2C3%5D%29


>And you conclude with this:

>>Ok, so we need to agree that add is "lazy" and uses the minimum arity.
That would remove any confusion.

> If this is you in the end reaching the conclusion that Ramda has done the right thing, that's fine. This might then make an interesting Gist or blog post, but perhaps not an issue for the Ramda list. If that's the case, please close this issue.

>If you do think there is something for Ramda to change, could you please state it succinctly?

Well before throwing a list of suggestions, I wanted to hear some feedback in case my understanding was wrong. But here you go:

Suggestions
---

1. Declare alias `add2` for `add`. That name reveals uniquely and unambigously the intent, and also alerts every user of `add` to the fact that `add` itself is binary. Which right now is easy to miss. 

2. Also consider the new function `add3` for the corresponding ternary version. 

3. Consider similar aliasing for other functions, wherever any confusion is possible.

4.? I am hesitating to suggest the fully variadic implementation idea I mentioned above, as it might not be in line with Ramda's design and might create additional complexity. At least not until there are clear use cases. Just keeping the idea for the record, that technically it might actually be possible, if needed.

**Comment by fix-fix** [comment]
@dmitriz addition is defined as a binary operation and function's arity in Ramda is already fixed, therefore there is no reason to specify its arity in its name.
There are also no reasons for functions like `addN`, since addition, unlike function composition, is commutative and you can just fold as many values as you want in any way.

In general it's not a good idea for a function to be variadic, since it breaks currying.

**Comment by CrossEye** [comment]
> Using generic name like add, which means variadic for most people, for something different, i.e. binary addition, is confusing and potentially dangerous.

Ah, but I disagree that it's confusing to most people.  In fact, as far as I recall, this is the first time anyone has made this objection.  `add` is simply the language's `+` operator, restricted (since about a year ago) to numbers, and made into a plain function.  If you want one that operates on a list of numbers, you can use [`sum`][su].


> Variadic functions are commonly used in JavaScript.
> And even in Ramda I can see `compose` and `pipe` :)

Those are some of the very few remaining variadic functions in Ramda, ones that are harder to remove for backwards-compatibility reasons, but which might well change to taking a list after 1.0.

Yes, variadic functions are commonly used in Javascript.  For Ramda, that means that we try to work with users' variadic functions where possible.  (This is a big difference from Sanctuary, which takes a stronger stance against them.)  But that does not imply that Ramda will expose variadic functions itself.  Variadic functions cannot be curried in any straightforward manner.  And currying is extremely important to the functional composition style central to Ramda's mission.


> **Suggestion**
>
> Use `add2` and `add3` instead of `add` for corresponding binary and ternary versions.

What about `add17` and `add42`?  Sanctuary's `lift` can easily be limited by saying that reasonable functions don't need more than a few parameters.  You can't say the same thing about addition.

>>> ```js
>>> // Possible changed behaviour - Any room for confusion?
>>> [1,2,3].map(add(3, 1)) / => [5,6,7]
>>> // currently error: '4 is not a function'
>>> ```

>> Well no, no room for confusion, but by the rules of Javascript it cannot work like you seem to want. Javascript is eagerly evaluated. So when you pass an expression to map, it must be evaluated before map is called. So the first step is to evaluated `add(3, 1)`, which will yield 4, and then call `[1, 2, 3].map(4)`, and hence the error you see. That expression cannot know its own context. You cannot say to `add(3, 1)`, "Hey, you're about to be called in a map, so please return a function rather than a scalar value."

> I see, but can I not evaluate `add(3,1)` to both number and function (stored in an object, and lazily for performance reasons)? Now `map(...)` comes along looking at this object and says "Hey, I can pick the function variant" and off you go ;) Intuitive for the user and no errors. Does it make sense?

This would mean altering every Ramda function that expects a function and making it test for an object with the magical property you suggest.  And ones that expect a value would also need such a test.  Worse would be distinguishing whether this object is itself a legitimate value or just such a wrapper.  But worst of all, a user of Ramda could no longer use `add(first, second)` normally.  She would have to check the type of the return, and then pull the appropriate value or function from the result, then either use the value or call the function.  This is far, far from the simplicity Ramda strives for.

 > [ ... ]
> Hm... I actually get this:
> ```js
> R.map(add(3), add(1))([1,2,3]) 
> // => "311,2,3"
> ```

I think you might still be missing some of the point. 

```js
R.map(add(3), add(1)); 
//~~> x => add(3, add(1, x)) ~~> 3 + 1 + x   // in version 0.19
// or
//                           ~~> Number(3) + Number(1) + Number(x) // in versions >= 0.20
```

(I'm not sure why you're testing with a year-old version of Ramda, but the behavior of `add` was updated a year ago to really work only on numbers.)

That `map` call in there happens to work as function composition.  You are not successfully mapping the resulting function over the array.  If you want that, you can do

```js
map(map(add(3), add(1)), [1,2,3]); //=> [5, 6, 7]
```

This breaks down something like

```js
map(map(add(3), add(1)), [1,2,3])
//~~> map(x => add(3, add(1, x)), [1, 2, 3])
//~~> map(x => 3 + 1 + x, [1, 2, 3])
//~~> [3 + 1 + 1, 3 + 1 +2, 3 + 1 + 3]
//~~> [5, 6, 7]
```

So, yes, that's my feedback.  I actually think this is quite wrong for Ramda.  `add` is reasonable as a binary function.  `sum` can serve for adding multiple elements.  But if we didn't include it, you could easily create your own version with

```js
const sum = reduce(add, 0)
```

Or, if you really want a variadic version, you could do

```js
const addAll = unapply(reduce(add, 0))
```

I'm sorry if I come across as harsh here.  This is further from the spirit of Ramda than most suggestions, and I'm not feeling well, so I may not be responding as charitably as usual.  If so, I apologize.

  [su]: http://ramdajs.com/docs/#sum

**Comment by paldepind** [comment]
The documentation for `add` clearly states:

> Adds two values.

And silently ignoring extra arguments is how JavaScript rolls.

I can see why some might think `add` would be variadic. But since the documentation is clear I see no problem. Ramda as a whole goes against what some JS developers would expect. It seems fair to expect that people read the documentation for the functions they're using.

**Comment by davidchambers** [comment]
Sanctuary functions do not permit unexpected arguments:

```javascript
S.add(2, 3, 10)
// ! TypeError: Function applied to too many arguments
//
//   add :: FiniteNumber -> FiniteNumber -> FiniteNumber
//
//   ‘add’ expected at most two arguments but received three arguments.
```

Ramda will not provide type checking, so this issue can be closed.
